src/Util.h

```diff
@@ -50,6 +51,25 @@
 namespace Halide {
 namespace Internal {
 
+/** Some numeric conversions are UB if the value won't fit in the result;
+ * this clamps the value to a legal range to avoid the UB. */
+template<typename DST, typename SRC>
+DST safe_numeric_cast(SRC s) {
+    if (std::is_integral<DST>::value &&
+        (std::is_floating_point<SRC>::value ||
+         (std::is_integral<SRC>::value && std::is_signed<DST>::value && sizeof(DST) < sizeof(SRC)))) {
+        // float -> int is UB if value won't fit
+        // any int -> signed int is UB if value won't fit
```

Comment:
Given this statement, Is it correct to check sizeof(DST) < sizeof(SRC)? Seems like casting uintX -> intX where the value is over the signed range might be UB?

And even if not technically UB, I wonder if "safe_numeric_cast" should simply always clamp the value to the range of the output type?

Reply:
Would a saturating_cast, which has no UB, serve the same need as safe_numeric_cast?

Reply:
> Is it correct to check sizeof(DST) < sizeof(SRC)?

Good point, this might not be correct for uint -> int cases; let me re-check.

> Would a saturating_cast

It wouldn't match existing C++ behavior for (eg) a narrowing uint -> uint cast, which is well-defined as (essentially) doing a bitwise and, not a clamp.